### PR TITLE
Rotation cleanup

### PR DESCRIPTION
**Pull request by RudolfWeeber**

### CODE REVIEW

**Comment by RudolfWeeber** [comment]
Fixes #2383.

The engine tests only check against a stored reference file.
I checked that the engine_lb test passes, if I re-introduce the sign error.
This PR replaced the reference data with data generated with the correct cross product. @hmenke @jdgraaf

File: src/core/communication.cpp [hunk 1]
```diff
@@ -815,7 +816,7 @@ void mpi_send_quat_slave(int pnode, int part) {
 
 /********************* REQ_SET_OMEGA ********/
 
-void mpi_send_omega(int pnode, int part, double omega[3]) {
+void mpi_send_omega(int pnode, int part, const double omega[3]) {
```
Comment (KaiSzuttor) [inline src/core/communication.cpp]: const Vector3d& omega

File: src/core/communication.cpp [hunk 2]
```diff
@@ -847,7 +848,7 @@ void mpi_send_omega_slave(int pnode, int part) {
 
 /********************* REQ_SET_TORQUE ********/
 
-void mpi_send_torque(int pnode, int part, double torque[3]) {
+void mpi_send_torque(int pnode, int part, const double torque[3]) {
```
Comment (KaiSzuttor) [inline src/core/communication.cpp]: const Vector3d& torque

File: src/core/rotation.cpp [hunk 3]
```diff
@@ -254,21 +254,45 @@ void propagate_omega_quat_particle(Particle *p) {
...
 
   ONEPART_TRACE(if (p->p.identity == check_id)
                     fprintf(stderr, "%d: OPT: PPOS p = (%.3f,%.3f,%.3f)\n",
                             this_node, p->r.p[0], p->r.p[1], p->r.p[2]));
 }
 
+inline
+
+    void
+    convert_torque_to_body_frame_apply_fix_and_thermostat(Particle &p) {
```
Comment (KaiSzuttor) [inline src/core/rotation.cpp]: if there is an “and” in the function name you should think about splitting it

**Review by KaiSzuttor — DISMISSED** [review]

File: src/core/communication.cpp [hunk 4]
```diff
@@ -644,15 +644,16 @@ void mpi_send_rotational_inertia_slave(int pnode, int part) {
...
 #ifdef ROTATION
   mpi_call(mpi_rotate_particle_slave, pnode, part);
 
   if (pnode == this_node) {
     Particle *p = local_particles[part];
-    local_rotate_particle(p, axis, angle);
+    local_rotate_particle(*p, axis, angle);
   } else {
-    MPI_Send(axis, 3, MPI_DOUBLE, pnode, SOME_TAG, comm_cart);
+    MPI_Send(axis.data(), 3, MPI_DOUBLE, pnode, SOME_TAG, comm_cart);
```
Comment (fweik) [inline src/core/communication.cpp]: Please use boost::mpi, which will also fix all the builds.

File: src/core/communication.cpp [hunk 5]
```diff
@@ -664,10 +665,12 @@ void mpi_rotate_particle_slave(int pnode, int part) {
 #ifdef ROTATION
   if (pnode == this_node) {
     Particle *p = local_particles[part];
-    double axis[3], angle;
-    MPI_Recv(axis, 3, MPI_DOUBLE, 0, SOME_TAG, comm_cart, MPI_STATUS_IGNORE);
+    Vector3d axis;
+    double angle;
+    MPI_Recv(axis.data(), 3, MPI_DOUBLE, 0, SOME_TAG, comm_cart,
```
Comment (fweik) [inline src/core/communication.cpp]: boost::mpi

File: src/core/minimize_energy.cpp [hunk 6]
```diff
@@ -95,7 +95,7 @@ bool steepest_descent_step(void) {
     }
 #ifdef ROTATION
     // Rotational increment
-    double dq[3]; // Vector parallel to torque
+    Vector3d dq; // Vector parallel to torque
```
Comment (fweik) [inline src/core/minimize_energy.cpp]: Looks like can directly initialize `dq` and remove the loop in line 100.

File: src/core/rotation.cpp [hunk 7]
```diff
@@ -254,21 +254,45 @@ void propagate_omega_quat_particle(Particle *p) {
...
-  // When dipoles are enabled, update dipole moment
-#endif
 
   ONEPART_TRACE(if (p->p.identity == check_id)
                     fprintf(stderr, "%d: OPT: PPOS p = (%.3f,%.3f,%.3f)\n",
                             this_node, p->r.p[0], p->r.p[1], p->r.p[2]));
 }
 
+inline
+
```
Comment (fweik) [inline src/core/rotation.cpp]: remove empty line

File: src/core/rotation.cpp [hunk 8]
```diff
@@ -490,66 +404,32 @@ Vector3d convert_vector_space_to_body(const Particle &p, const Vector3d &v) {
...
-
-  res[0] = A[0 + 3 * 0] * v[0] + A[0 + 3 * 1] * v[1] + A[0 + 3 * 2] * v[2];
-  res[1] = A[1 + 3 * 0] * v[0] + A[1 + 3 * 1] * v[1] + A[1 + 3 * 2] * v[2];
-  res[2] = A[2 + 3 * 0] * v[0] + A[2 + 3 * 1] * v[1] + A[2 + 3 * 2] * v[2];
-}
-
 /** Rotate the particle p around the NORMALIZED axis aSpaceFrame by amount phi
  */
-void local_rotate_particle(Particle *p, double *aSpaceFrame, double phi) {
+void local_rotate_particle(Particle &p, const Vector3d &a_space_frame,
```
Comment (fweik) [inline src/core/rotation.cpp]: Please rename `a_space_frame` to `axis`.

**Review by fweik — CHANGES_REQUESTED** [review]

**Comment by RudolfWeeber** [comment]
Fixes #1318

**Review by fweik — APPROVED** [review]

**Pull request merged**